File: LeetCode/1137-n-th-tribonacci-number.py
###############################################################################################
# 简单
###########
# 时间复杂度：O(n)
# 空间复杂度：O(1)
###############################################################################################
class Solution:
    def tribonacci(self, n: int) -> int:
        if n < 3:
            return n if n != 2 else 1
        f0, f1, f2 = 0, 1, 1
        for _ in range(3, n+1):
            f3 = f0 + f1 + f2
            f2, f1, f0 = f3, f2, f1
        return f3

# 递归写法
class Solution:
    def tribonacci(self, n: int) -> int:
        if n < 3:
            return n if n != 2 else 1
        def dfs(f0, f1, f2, l):
            if l > n:
                return f2
            return dfs(f1, f2, f0+f1+f2, l+1)
        return dfs(0, 1, 1, 3)

File: LeetCode/1137-n-th-tribonacci-number.cpp
///////////////////////////////////////////////////////////////////////////////////////////////
// easy
////////////////
// Time Complexity: O(n)
// Space Complexity: O(n)
///////////////////////////////////////////////////////////////////////////////////////////////
class Solution {
public:
    int v[40];
    int tribonacci(int n) {
        if (n == 0) return 0;
        if (n == 1 || n == 2) return 1;
        if (v[n] != 0) return v[n];
        else v[n] = tribonacci(n-1) + tribonacci(n-2) + tribonacci(n-3);
        return v[n];
    }
};